lib/queue/queue.go

```diff
@@ -0,0 +1,136 @@
+package queue
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/google/uuid"
+	"github.com/voxpupuli/webhook-go/config"
+	"github.com/voxpupuli/webhook-go/lib/helpers"
+)
+
+type Queue struct {
+	Items   []*QueueItem
+	wg      sync.WaitGroup
+	jobChan chan *QueueItem
+}
+
+type QueueItem struct {
+	Id          uuid.UUID
+	Name        string
+	CommandType string
+	AddedAt     time.Time
+	StartedAt   time.Time
+	FinishedAt  time.Time
+	Command     []string
+	Response    interface{}
+	State       string
+}
+
+var queue = &Queue{}
+
+func init() {
+	conf := config.GetConfig()
+
+	log.Printf("start queue with %d jobs", conf.Server.Queue.MaxConcurrentJobs)
+	queue.jobChan = make(chan *QueueItem, 2)
+	queue.wg.Add(1)
+}
+
+func GetQueueItems() []*QueueItem {
+	return queue.Items
+}
+
+func AddToQueue(commandType string, name string, command []string) (*QueueItem, error) {
+	id, err := uuid.NewUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	queueItem := QueueItem{
+		Id:          id,
+		AddedAt:     time.Now(),
+		Command:     command,
+		CommandType: commandType,
+		Name:        name,
+		State:       "added",
+	}
+
+	queue.Items = append(queue.Items, &queueItem)
+
+	if !queueJob(&queueItem) {
+		queueItem.State = "queue full"
+		return &queueItem, fmt.Errorf("queue is full")
+	}
+
+	return &queueItem, nil
+}
+
+func trimItems() {
+	conf := config.GetConfig()
+
+	if conf.Server.Queue.MaxHistoryItems == 0 {
+		return
+	}
+
+	if len(queue.Items) > conf.Server.Queue.MaxHistoryItems {
+		queue.Items = queue.Items[:conf.Server.Queue.MaxHistoryItems]
+	}
+}
+
+func Work() error {
+	go worker()
+	return nil
+}
+
+func Dispose() {
+	close(queue.jobChan)
+}
+
+func queueJob(command *QueueItem) bool {
+	select {
+	case queue.jobChan <- command:
+		return true
+	default:
+		return false
+	}
+}
+
+func worker() {
+	defer queue.wg.Done()
+
+	log.Println("Worker is waiting for jobs")
+
+	conf := config.GetConfig()
+	conn := helpers.ChatopsSetup()
+
+	for job := range queue.jobChan {
+		log.Println("Worker picked Job", job.Id)
+		job.StartedAt = time.Now()
+		job.State = "started"
```

Comment:
Would setting this to `"running"` be a better state than `"started"`? I'm impartial, but since you are already tracking the start time in the `StartedAt` field, I feel like a `"started"` state is redundant